lib/TPP/ConvertPackUnpackOptimization.cpp

```diff
@@ -0,0 +1,216 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+
+#include "TPP/BuilderUtils.h"
+#include "TPP/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+using namespace mlir::tpp;
+
+#define GEN_PASS_CLASSES
+#include "TPP/Passes.h.inc"
+
+namespace {
+
+template <typename Op, bool IsPack>
+class ConvertPackUnpackOptimizationOp : public OpRewritePattern<Op> {
+  using OpRewritePattern<Op>::OpRewritePattern;
+  LogicalResult matchAndRewrite(Op packOp,
+                                PatternRewriter &rewriter) const override {
+    if (packOp.getStaticInnerTiles().size() <= 0)
+      return failure();
+    ArrayRef<int64_t> shape = IsPack ? packOp.getSourceType().getShape()
+                                     : packOp.getDestType().getShape();
+    int numLoops = shape.size();
+
+    for (size_t i = 0; i < packOp.getInnerDimsPos().size() - 1; i++) {
+      if (packOp.getInnerDimsPos()[i] >= packOp.getInnerDimsPos()[i + 1]) {
+        return failure();
+      }
+    }
+
+    for (size_t i = 0; i < packOp.getStaticInnerTiles().size(); i++) {
+      if (shape[packOp.getInnerDimsPos()[i]] %
+              packOp.getStaticInnerTiles()[i] !=
+          0) {
+        return failure();
+      }
+    }
+    auto zero = getConstIndex(rewriter, 0);
+    auto one = getConstIndex(rewriter, 1);
+
+    SmallVector<Value> lbs(numLoops, zero);
+    SmallVector<Value> ubs;
+    SmallVector<Value> steps(numLoops, one);
+
+    std::map<int, int> tiledDims;
+    for (auto tiledDim : llvm::enumerate(packOp.getInnerDimsPos())) {
+      tiledDims[tiledDim.value()] = tiledDim.index();
+    }
+
+    for (int i = 0; i < numLoops; i++) {
+      if (tiledDims.count(i) && shape[i] != ShapedType::kDynamic &&
+          packOp.getStaticInnerTiles()[tiledDims[i]] != ShapedType::kDynamic) {
+        ubs.push_back(getConstIndex(
+            rewriter, shape[i] / packOp.getStaticInnerTiles()[tiledDims[i]]));
+
+      } else {
+        ubs.push_back(getConstIndex(rewriter, shape[i]));
+      }
+    }
+
+    auto loopNest = mlir::scf::buildLoopNest(
+        rewriter, packOp.getLoc(), lbs, ubs, steps, packOp.getDest(),
+        [&packOp, &numLoops,
+         &tiledDims](OpBuilder &rewriter, Location loc, ValueRange localIvs,
+                     ValueRange iterArgs) -> scf::ValueVector {
+          SmallVector<OpFoldResult> extractSliceOffsets;
+          SmallVector<OpFoldResult> extractSliceStrides;
+          SmallVector<OpFoldResult> extractSliceSizes;
+          SmallVector<OpFoldResult> insertSliceOffsets;
+          SmallVector<OpFoldResult> insertSliceStrides;
+          SmallVector<OpFoldResult> insertSliceSizes;
+
+          if (IsPack) {
```

Comment:
I can't see the difference between `IsPack` and not, but it doesn't look big. Can you extract the differences and have a single block of code for both?

Reply:
Problem is offsets, strides and sizes are interchanged for extract and insert slice operations in case of pack and unpack operations. So it's a little tricky to refactor this into common code of some kind. @alheinecke 
is of the opinion that we can refactor it after merging when we get to it while working on vnni pack and unpack.